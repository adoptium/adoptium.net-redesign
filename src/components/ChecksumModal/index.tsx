import React, { Fragment, useRef, useState, useEffect } from "react"
import { Dialog, Transition } from "@headlessui/react"
import { FaClipboard } from "react-icons/fa"

const ChecksumModal = ({ open, setOpen, checksum }) => {
  const cancelButtonRef = useRef(null)
  const [copied, setCopied] = useState(false)

  useEffect(() => {
    let timer
    if (copied) {
      timer = setTimeout(() => {
        setCopied(false)
      }, 3000)
    }

    return () => clearTimeout(timer)
  }, [copied])

  return (
    <Transition.Root show={open} as={Fragment}>
      <Dialog
        as="div"
        className="relative z-10"
        initialFocus={cancelButtonRef}
        onClose={() => setOpen(false)}
      >
        <Transition.Child
          as={Fragment}
          enter="ease-out duration-300"
          enterFrom="opacity-0"
          enterTo="opacity-100"
          leave="ease-in duration-200"
          leaveFrom="opacity-100"
          leaveTo="opacity-0"
        >
          <div className="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity" />
        </Transition.Child>

        <div className="fixed inset-0 z-10 overflow-y-auto">
          <div className="flex min-h-full items-end justify-center p-4 text-center sm:items-center sm:p-0">
            <Transition.Child
              as={Fragment}
              enter="ease-out duration-300"
              enterFrom="opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95"
              enterTo="opacity-100 translate-y-0 sm:scale-100"
              leave="ease-in duration-200"
              leaveFrom="opacity-100 translate-y-0 sm:scale-100"
              leaveTo="opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95"
            >
              <Dialog.Panel className="relative transform overflow-hidden rounded-lg bg-white text-left shadow-xl transition-all sm:my-8 sm:w-full sm:max-w-2xl">
                <div className="bg-[#200E46] px-4 pb-4 pt-5 sm:p-6 sm:pb-4">
                  <div className="sm:flex sm:items-start">
                    <div className="mx-auto flex h-12 w-12 flex-shrink-0 items-center justify-center rounded-full bg-blue-100 sm:mx-0 sm:h-10 sm:w-10">
                      <FaClipboard
                        className="h-6 w-6 text-blue-600"
                        aria-hidden="true"
                      />
                    </div>
                    <div className="mt-3 text-center sm:mt-0 sm:ml-4 sm:text-left">
                      <Dialog.Title
                        as="h3"
                        className="text-lg leading-6 font-medium text-white"
                      >
                        Checksum (SHA256)
                      </Dialog.Title>
                      <div className="mt-2">
                        <p className="text-sm text-white">
                          The checksum is used to ensure the file has not been
                          corrupted during download. You should compare the
                          checksum of the file you received with the value below
                          to ensure the file you received is complete and
                          unmodified.
                        </p>
                        <p className="text-sm text-white mt-2">
                          On Windows use{" "}
                          <code>certUtil -hashfile file SHA256</code>, <br /> on
                          Linux use <code>sha256sum file</code>, <br /> and on
                          macOS use <code>shasum -a 256 file</code>.
                        </p>
                        <input
                          readOnly
                          className="mt-2 w-full text-gray-900 rounded-md border-gray-300 shadow-sm"
                          value={checksum}
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div className="bg-purple px-4 py-3 sm:flex sm:flex-row-reverse sm:px-6">
                  <button
                    type="button"
                    className={`mt-3 w-full inline-flex justify-center rounded-md border border-transparent shadow-sm px-4 py-2 text-base font-medium text-white sm:mt-0 sm:ml-3 sm:w-auto sm:text-sm ${
                      copied ? "bg-[#2B1A4F]" : "bg-pink hover:bg-pink-700"
                    }`}
                    onClick={() => {
                      navigator.clipboard.writeText(checksum)
                      setCopied(true)
                    }}
                  >
                    {copied ? "Copied" : "Copy"}
                  </button>
                  <button
                    type="button"
                    className="mt-3 w-full inline-flex justify-center rounded-md border border-gray-300 shadow-sm px-4 py-2 bg-white text-base font-medium text-gray-700 hover:bg-gray-50 sm:mt-0 sm:w-auto sm:text-sm"
                    onClick={() => setOpen(false)}
                    ref={cancelButtonRef}
                  >
                    Close
                  </button>
                </div>
              </Dialog.Panel>
            </Transition.Child>
          </div>
        </div>
      </Dialog>
    </Transition.Root>
  )
}

export default ChecksumModal
