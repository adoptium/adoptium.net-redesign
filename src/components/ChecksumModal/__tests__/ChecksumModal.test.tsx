import React from "react"
import userEvent from "@testing-library/user-event"
import { render, screen, act } from "@testing-library/react"
import { describe, expect, it, vi } from "vitest"
import ChecksumModal from ".." // Ensure this import path is correct

// Setup for navigator.clipboard mock
Object.assign(navigator, {
  clipboard: {
    writeText: vi.fn(),
  },
})

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: vi.fn().mockImplementation(() => ({
    matches: false,
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
  })),
})

window.ResizeObserver = vi.fn().mockImplementation(() => ({
  observe: vi.fn(),
  unobserve: vi.fn(),
  disconnect: vi.fn(),
}))

const navigatorClipboardSpy = vi.spyOn(navigator.clipboard, "writeText")

describe("ChecksumModal component", () => {
  // Providing the required props for the component to render correctly
  const requiredProps = {
    open: true,
    setOpen: vi.fn(), // Mocking setOpen function
    checksum: "exampleChecksumValue",
  }

  it("ChecksumModal renders correctly", () => {
    const { container } = render(
      <ChecksumModal
        open={true}
        setOpen={() => {}}
        checksum="exampleChecksumValue"
      />,
    )
    // wait for the component to be in the document
    expect(screen.getByText("Checksum (SHA256)")).toBeInTheDocument()
    // Find the input field by its role and ensure it's in the document
    const inputField = screen.getByRole("textbox")
    expect(inputField).toBeInTheDocument()

    // Assert that the input field has the correct value
    expect(inputField).toHaveValue("exampleChecksumValue")
    expect(container).toMatchSnapshot()
  })

  it("ChecksumModal copies correctly", async () => {
    render(<ChecksumModal {...requiredProps} />)
    navigatorClipboardSpy.mockImplementationOnce(() => Promise.resolve())

    // Using act to wait for all state updates and effects to finish
    await act(async () => {
      userEvent.click(screen.getByText("Copy"))
    })

    // Since the text changes to "Copied" after the button is clicked and state is updated,
    // we need to ensure the component has re-rendered with the new state
    await screen.findByText("Copied")

    expect(navigatorClipboardSpy).toHaveBeenCalledTimes(1)
    expect(navigatorClipboardSpy).toHaveBeenCalledWith("exampleChecksumValue")
  })
})
