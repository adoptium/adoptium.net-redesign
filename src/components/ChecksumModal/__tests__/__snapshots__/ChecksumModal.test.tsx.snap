// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChecksumModal component > ChecksumModal renders correctly 1`] = `
<div
  aria-hidden="true"
>
  <div
    hidden=""
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;
